src/box/memtx_engine.cc

```diff
@@ -1265,6 +1265,84 @@ struct memtx_join_ctx {
 	struct xstream *stream;
 };
 
+static void
+send_join_meta(struct xstream *stream, const struct vclock *vclock,
+	       const bool send_meta, const struct raft_request *raft_req,
+	       const struct synchro_request *synchro_req)
+{
+	/* Respond to the JOIN request with the current vclock. */
+	struct xrow_header row;
+	/* Encoding replication request uses fiber()->gc region. */
+	RegionGuard region_guard(&fiber()->gc);
+	/*
+	 * Vclock is encoded with 0th component, as in case of checkpoint
+	 * join it corresponds to the vclock of the checkpoint, where 0th
+	 * component is essential, as otherwise signature won't be correct.
+	 * Client sends this vclock in IPROTO_CURSOR, when he wants to
+	 * continue fetching from the same checkpoint.
+	 */
+	xrow_encode_vclock(&row, vclock);
+	xstream_write(stream, &row);
+
+	/*
+	 * Version is present starting with 2.7.3, 2.8.2, 2.9.1
+	 * All these versions know of additional META stage of initial join.
+	 */
+	if (send_meta) {
+		/* Mark the beginning of the metadata stream. */
+		xrow_encode_type(&row, IPROTO_JOIN_META);
+		xstream_write(stream, &row);
+
+		xrow_encode_raft(&row, &fiber()->gc, raft_req);
+		xstream_write(stream, &row);
+
+		char body[XROW_BODY_LEN_MAX];
+		xrow_encode_synchro(&row, body, synchro_req);
+		row.replica_id = synchro_req->replica_id;
+		xstream_write(stream, &row);
+
+		/* Mark the end of the metadata stream. */
+		xrow_encode_type(&row, IPROTO_JOIN_SNAPSHOT);
+		xstream_write(stream, &row);
+	}
+}
+
+#if defined(ENABLE_FETCH_SNAPSHOT_CHECKPOINT_CURSOR)
+#include "memtx_checkpoint_join.c"
+#else /* !defined(ENABLE_FETCH_SNAPSHOT_CHECKPOINT_CURSOR) */
+
+static int
+memtx_engine_prepare_checkpoint_join(struct engine *engine,
+				     struct engine_join_ctx *ctx)
+{
+	(void)engine;
+	(void)ctx;
+	diag_set(ClientError, ER_UNSUPPORTED, "Tarantool CE",
+		 "checkpoint join");
```

Comment:
Nit: Usually we say that "Community edition does not support <feature>", not "Tarantool CE":

https://github.com/tarantool/tarantool/blob/319357d5973d15d08b8eda6a230eada08b710802/src/box/lua/schema.lua#L924

https://github.com/tarantool/tarantool/blob/319357d5973d15d08b8eda6a230eada08b710802/src/box/lua/schema.lua#L4090

Reply:
Indeed, fixed